src/ess/reduce/time_of_flight/toa_to_tof.py

```diff
@@ -111,231 +104,289 @@ def compute_tof_lookup_table(
+    new_bins = sc.sort(
+        sc.concat(
+            [da.coords[key], sc.scalar(0.0, unit=eto_unit), pulse_period], dim=key
+        ),
+        key=key,
+    )
+    rebinned = da.rebin({key: new_bins})
+    etos = rebinned.coords[key]
+
+    # In histogram mode, the lookup table cannot have a pulse dimension because we
+    # cannot know in the histogrammed data which pulse the events belong to.
+    # So we merge the pulse dimension in the lookup table. A quick way to do this
+    # is to take the mean of the data along the pulse dimension (there should
+    # only be regions that are NaN in one pulse and finite in the other).
+    merged = lookup.data.nanmean('pulse')
+    dim = merged.dims[0]
+    lookup = sc.DataArray(
+        data=merged.fold(dim=dim, sizes={'pulse': 1, dim: merged.sizes[dim]}),
+        coords={
+            'pulse': sc.arange('pulse', 1.0),
+            'distance': lookup.coords['distance'],
+            'event_time_offset': lookup.coords['event_time_offset'],
+        },
+    )
+    pulse_index = sc.zeros(sizes=etos.sizes)
 
-    Parameters
-    ----------
-    toa:
-        Time of arrival of the neutron at the detector, unwrapped at the pulse period.
-    pivot_time:
-        Pivot time at the detector, i.e., the time of the start of the frame at the
-        detector.
-    """
-    # Order of operation to preserve dimension order
-    return UnwrappedTimeOfArrivalMinusPivotTime(
-        -pivot_time.to(unit=elem_unit(toa), copy=False) + toa
+    # Create 2D interpolator
+    interp = _make_tof_interpolator(
+        lookup, distance_unit=ltotal.unit, time_unit=eto_unit
     )
 
+    # Compute time-of-flight of the bin edges using the interpolator
+    tofs = sc.array(
+        dims=etos.dims,
+        values=interp((pulse_index.values, ltotal.values, etos.values)),
+        unit=eto_unit,
+    )
 
-def time_of_arrival_minus_pivot_time_modulo_period(
-    toa_minus_pivot_time: UnwrappedTimeOfArrivalMinusPivotTime,
-    frame_period: FramePeriod,
-) -> TimeOfArrivalMinusPivotTimeModuloPeriod:
-    """
-    Compute the time of arrival of the neutron at the detector, unwrapped at the pulse
-    period, minus the start time of the frame, modulo the frame period.
+    return rebinned.assign_coords(tof=tofs)
+
+
+def _time_of_flight_data_events(
+    da: sc.DataArray,
+    lookup: sc.DataArray,
+    ltotal: sc.Variable,
+    pulse_period: sc.Variable,
+    pulse_stride: int,
+) -> sc.DataArray:
+    etos = da.bins.coords["event_time_offset"]
+    eto_unit = elem_unit(etos)
+    pulse_period = pulse_period.to(unit=eto_unit)
+    frame_period = pulse_period * pulse_stride
+
+    # Compute a pulse index for every event: it is the index of the pulse within a
+    # frame period. When there is no pulse skipping, those are all zero. When there is
+    # pulse skipping, the index ranges from zero to pulse_stride - 1.
+    etz = da.bins.concat().value.coords['event_time_zero']
+    tmin = etz.min()
```

Comment:
Can't tell were `etz` is used. Isn't `tmin` just
```suggestion
    tmin = da.bins.coords['event_time_zero'].min()
```

Reply:
Left over from a refactor I think :+1: